refactor(compiler): introduce directive wrappers to generate less code

tbosch opened the pull request.

tbosch committed:
refactor(compiler): introduce directive wrappers to generate less code

- for now only wraps the `@Input` properties and calls
  to `ngOnInit`, `ngDoCheck` and `ngOnChanges` of directives.
- also groups eval sources by NgModule.

Part of #11683

vicb commented on modules/@angular/compiler/src/directive_wrapper_compiler.ts:
```diff
@@ -0,0 +1,180 @@
+const RESET_CHANGES_STMT = o.THIS_EXPR.prop(CHANGES_FIELD_NAME).set(o.literalMap([])).toStmt();
+
+@Injectable()
+export class DirectiveWrapperCompiler {
```
inline comments

tbosch replied:
Done

vicb commented on modules/@angular/compiler/src/offline_compiler.ts:
```diff
@@ -69,8 +69,8 @@ export class OfflineCompiler {
-      moduleUrl: string, ngModulesSummary: NgModulesSummary, components: StaticSymbol[],
-      ngModules: StaticSymbol[]): Promise<SourceModule[]> {
+      moduleUrl: string, ngModulesSummary: NgModulesSummary, directives: StaticSymbol[],
+      components: StaticSymbol[], ngModules: StaticSymbol[]): Promise<SourceModule[]> {
```
merge dir & cmp ?

tbosch replied:
Done

vicb commented on modules/@angular/compiler/src/offline_compiler.ts:
```diff
@@ -80,11 +80,15 @@ export class OfflineCompiler {
 
+    // compile directive wrappers
+    exportedVars.push(...directives.map(
+        (directiveType) => { return this._compileDirectiveWrapper(directiveType, statements); }));
```
nit: `{ return`

tbosch replied:
Done

vicb commented on modules/@angular/compiler/src/view_compiler/property_binder.ts:
```diff
@@ -230,81 +248,46 @@ export function bindDirectiveHostProps(
 
 export function bindDirectiveInputs(
     directiveAst: DirectiveAst, directiveInstance: o.Expression, compileElement: CompileElement) {
```
rename `directiveInstance` to `directiveWrapperInstance`

vicb commented on modules/@angular/compiler/src/directive_wrapper_compiler.ts:
```diff
@@ -0,0 +1,180 @@
+    const fields: o.ClassField[] = [
+      new o.ClassField(CONTEXT_FIELD_NAME, o.importType(dirMeta.type)),
+    ];
+    if (lifecycleHooks.genChanges) {
```
Merge with previous `if (genChanges)`

vicb commented on modules/@angular/compiler/src/directive_wrapper_compiler.ts:
```diff
@@ -0,0 +1,180 @@
+      fields.push(new o.ClassField(CHANGES_FIELD_NAME, new o.MapType(o.DYNAMIC_TYPE)));
+    }
+    const methods: o.ClassMethod[] = [];
+    const inputFieldNames: string[] = Object.keys(dirMeta.inputs);
```
Inline

tbosch replied:
Done

vicb commented on modules/@angular/compiler/src/directive_wrapper_compiler.ts:
```diff
@@ -0,0 +1,180 @@
+                                         .instantiate([fieldExpr, CURR_VALUE_VAR]))
+                                .toStmt());
+  }
+  onChangeStatements.push(fieldExpr.set(CURR_VALUE_VAR).toStmt());
```
Move up

tbosch replied:
Can't as we need to keep the old value

vicb commented on modules/@angular/compiler/src/directive_wrapper_compiler.ts:
```diff
@@ -0,0 +1,180 @@
+  }
+  onChangeStatements.push(fieldExpr.set(CURR_VALUE_VAR).toStmt());
+
+  const changedVar = o.variable('changed');
```
Remove variable.

tbosch replied:
Done

tbosch commented:
Verify performance impact on tree and table (create, update, destroy), should not have any.

tbosch committed:
refactor(compiler): remove private exports

All of `@angular/compiler` is private, so we can export
everything we need directly.

tbosch committed:
refactor(compiler): don’t use the `OfflineCompiler` in extract_i18n

tbosch commented on modules/@angular/compiler/src/identifiers.ts:
```diff
@@ -259,12 +262,22 @@ export class Identifiers {
+  };
+  static setBindingDebugInfo: IdentifierSpec = {
+    name: 'setBindingDebugInfo',
+    moduleUrl: assetUrl('core', 'i18n/tokens'),
```
Done

tbosch commented on modules/@angular/compiler/src/view_compiler/property_binder.ts:
```diff
@@ -31,15 +31,18 @@ function createCurrValueExpr(exprIndex: number): o.ReadVarExpr {
+}
+
+function evalCdAst(
+    view: CompileView, currValExpr: o.ReadVarExpr, parsedExpression: cdAst.AST,
```
I keep it for now

tbosch commented:
Checked the numbers on the tree benchmark, no regression for `create` nor `update`

tbosch commented:
Reviewed in person with @vicb

tbosch commented:
Note: The failure is due to our browserstack license.

alxhub commented:
This is failing in g3sync, likely due to build changes that are needed on the g3 side. I'll work with @alexeagle on getting this in.

alxhub commented:
Merged as b0a03fc.
